pkg/sql/catalog/hydratedtables/hydratedcache.go

```diff
@@ -0,0 +1,265 @@
+	// changes as different callers oscillate evicting the old and new versions of
+	// that type. It should converge rapidly but nevertheless, a finer-granularity
+	// cache could mitigate the problem.
+	key := makeIDVersion(table)
+	var groupKey string // used as a proxy for cache hit
+	defer func() {
+		if err != nil {
+			return
+		}
+		if groupKey == "" {
+			c.metrics.Hits.Inc(1)
+		} else {
+			c.metrics.Misses.Inc(1)
+		}
+	}()
+	for {
+		cached, ok := c.getHydratedTableDescriptorFromCache(ctx, key)
+		if ok {
+			canUse, err := canUseCachedDescriptor(ctx, cached, lookupFunc)
+			if err != nil {
+				return nil, err
+			}
+			if canUse {
+				return cached.tableDesc, nil
+			}
+		}
+		if groupKey == "" {
+			groupKey = fmt.Sprintf("%d@%d", key.ID, key.Version)
+		}
+		// Now we want to lock the cache and populate the descriptors.
+		// Only the calling goroutine can actually use the result directly.
+		// Furthermore, if an error is encountered, only the calling goroutine
+		// should return it.
+		var called bool
+		res, _, err := c.g.Do(groupKey, func() (interface{}, error) {
+			called = true
+			lookup := cachedTypeDescriptorResolver{
+				underlying: lookupFunc,
+				cache:      map[descpb.ID]*cachedType{},
+			}
+			descBase := protoutil.Clone(table.TableDesc()).(*descpb.TableDescriptor)
+			if err := typedesc.HydrateTypesInTableDescriptor(ctx, descBase, &lookup); err != nil {
+				return nil, err
+			}
+			hydrated := tabledesc.NewImmutable(*descBase)
+			types := make([]*cachedType, 0, len(lookup.cache))
+			for _, typ := range lookup.cache {
+				types = append(types, typ)
+			}
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			c.mu.cache.Add(key, &hydratedTableDescriptor{
+				tableDesc: hydrated,
+				typeDescs: types,
+			})
+			return hydrated, nil
+		})
+		if !called {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		return res.(*tabledesc.Immutable), nil
+	}
+}
+
+type cachedTypeDescriptorResolver struct {
+	underlying catalog.TypeDescriptorResolver
+	cache      map[descpb.ID]*cachedType
+}
+
+func (c cachedTypeDescriptorResolver) GetTypeDescriptor(
+	ctx context.Context, id descpb.ID,
+) (tree.TypeName, catalog.TypeDescriptor, error) {
+	if typ, exists := c.cache[id]; exists {
+		return typ.name, typ.typDesc, nil
+	}
+	name, typDesc, err := c.underlying.GetTypeDescriptor(ctx, id)
+	if err != nil {
+		return tree.TypeName{}, nil, err
+	}
+	c.cache[id] = &cachedType{
+		name:    name,
+		typDesc: typDesc,
+	}
+	return name, typDesc, nil
+}
+
+func canUseCachedDescriptor(
+	ctx context.Context, cached *hydratedTableDescriptor, res catalog.TypeDescriptorResolver,
+) (canUse bool, _ error) {
+	for _, typ := range cached.typeDescs {
+		name, typDesc, err := res.GetTypeDescriptor(ctx, typ.typDesc.GetID())
+		if err != nil {
+			return false, err
+		}
+		// Use raw pointer equality to check on the type descriptor being valid.
+		if typ.typDesc != typDesc ||
```

Comment:
Done. Pointer equality worked in my head because we get the same pointers out of the lease manager but I think you're spot on that this will spell trouble for the distsql flows.